pennylane/devices/null_qubit.py

```diff
@@ -11,320 +11,393 @@
-        return [0.0]
-
-    def _apply_t(self, state, axes, inverse=False):
-        return [0.0]
-
-    def _apply_sx(self, state, axes, inverse=False):
-        return [0.0]
-
-    def _apply_cnot(self, state, axes, **kwargs):
-        return [0.0]
-
-    def _apply_swap(self, state, axes, **kwargs):
-        return [0.0]
-
-    def _apply_cz(self, state, axes, **kwargs):
-        return [0.0]
-
-    def _apply_toffoli(self, state, axes, **kwargs):
-        return [0.0]
-
-    def _apply_phase(self, state, axes, parameters, inverse=False):
-        return [0.0]
-
-    def expval(self, observable, shot_range=None, bin_size=None):
-        return [0.0]
-
-    def var(self, observable, shot_range=None, bin_size=None):
-        return [0.0]
-
-    @classmethod
-    def capabilities(cls):
-        capabilities = super().capabilities().copy()
-        capabilities.update(
-            model="qubit",
-            supports_inverse_operations=True,
-            supports_analytic_computation=True,
-            supports_broadcasting=False,
-            returns_state=True,
-            passthru_devices={
-                "tf": "null.qubit",
-                "torch": "null.qubit",
-                "autograd": "null.qubit",
-                "jax": "null.qubit",
-            },
+    @staticmethod
+    def _derivatives(circuit, interface):
+        derivatives = ((math.asarray(0.0, like=interface),) * len(circuit.trainable_params)) * len(
+            circuit.measurements
         )
-        return capabilities
+        return derivatives[0] if len(derivatives) == 1 else derivatives
 
     @staticmethod
-    def _create_basis_state(index):
-        return [0.0]
+    def _vjp(circuit, interface):
+        return (math.asarray(0.0, like=interface),) * len(circuit.trainable_params)
 
-    @property
-    def state(self):
-        return [0.0]
+    @staticmethod
+    def _jvp(circuit, interface):
+        jvps = (math.asarray(0.0, like=interface),) * len(circuit.measurements)
+        return jvps[0] if len(jvps) == 1 else jvps
+
+    def preprocess(
+        self, execution_config=DefaultExecutionConfig
+    ) -> Tuple[TransformProgram, ExecutionConfig]:
+        updated_values = {}
+        if execution_config.gradient_method == "best":
+            updated_values["gradient_method"] = "device"
+        if execution_config.use_device_gradient is None:
+            updated_values["use_device_gradient"] = execution_config.gradient_method in {
+                "best",
+                "device",
+                "backprop",
+            }
+        if execution_config.use_device_jacobian_product is None:
+            updated_values["use_device_jacobian_product"] = (
+                execution_config.gradient_method == "device"
+            )
+        if execution_config.grad_on_execution is None:
+            updated_values["grad_on_execution"] = execution_config.gradient_method == "device"
+        return TransformProgram(), replace(execution_config, **updated_values)
```

Comment:
Potentially, we could create a target `default.qubit` device as just inherit the preprocessing from it.  That way we would still have "device like" preprocessing like decomposing templates.

Reply:
```
self._device = DefaultQubit(wires=self.wires, shots=self.shots)
return self._device.preprocess(config)
```

Reply:
I'm open to it, but now I should ask - do we want to tie `null.qubit` to `default.qubit` in general? I was trying to avoid it, although DQ is the de-facto standard simulator so I am not against it given that's where our maintenance efforts go

Reply:
I was thinking later on we could have something like:
```
qml.device('null.qubit', target=some_other_device)
```

Reply:
updated to use DQ's program, but still tweaking the config to support both "device" and "adjoint" to mean "device"